duperemove.8
.\" Automatically generated by Pandoc 3.1.11.1
.\"
.TH "Duperemove" "8" "01 Mar 2025" "duperemove 0.15.1" "System Manager\[cq]s Manual"
.SH NAME
\f[CR]duperemove\f[R] \- Find duplicate regions in files and submit them
for deduplication
.SH SYNOPSIS
\f[B]duperemove\f[R] \f[I]options\f[R] \f[I]files\&...\f[R]
.SH DESCRIPTION
\f[CR]duperemove\f[R] is a simple tool for finding duplicated regions in
files and submitting them for deduplication.
When given a list of files it will hash their contents and compare those
hashes to each other, finding and categorizing regions that match each
other.
When given the \f[CR]\-d\f[R] option, \f[CR]duperemove\f[R] will submit
those regions for deduplication using the Linux kernel FIDEDUPERANGE
ioctl.
.PP
\f[CR]duperemove\f[R] computes hashes for each files extents as well as
for the whole file\[cq]s content.
Optionally, per\-block hashes can be computed.
.PP
\f[CR]duperemove\f[R] can store the hashes it computes in a
\f[CR]hashfile\f[R].
If given an existing hashfile, \f[CR]duperemove\f[R] will only compute
hashes for those files which have changed since the last run.
Thus you can run \f[CR]duperemove\f[R] repeatedly on your data as it
changes, without having to re\-checksum unchanged data.
For more on hashfiles see the \f[CR]\-\-hashfile\f[R] option below as
well as the \f[CR]Examples\f[R] section.
.PP
\f[CR]duperemove\f[R] can also take input from the \f[CR]fdupes\f[R]
program, see the \f[CR]\-\-fdupes\f[R] option below.
.SH GENERAL
Duperemove has two major modes of operation, one of which is a subset of
the other.
.SS Readonly / Non\-deduplicating Mode
When run without \f[CR]\-d\f[R] (the default) duperemove will print out
one or more tables of matching extents it has determined would be ideal
candidates for deduplication.
As a result, readonly mode is useful for seeing what duperemove might do
when run with \f[CR]\-d\f[R].
.PP
Generally, duperemove does not concern itself with the underlying
representation of the extents it processes.
Some of them could be compressed, undergoing I/O, or even have already
been deduplicated.
In dedupe mode, the kernel handles those details and therefore we try
not to replicate that work.
.SS Deduping Mode
This functions similarly to readonly mode with the exception that the
duplicated extents found in our \[lq]read, hash, and compare\[rq] step
will actually be submitted for deduplication.
Extents that have already been deduped will be skipped.
An estimate of the total data deduplicated will be printed after the
operation is complete.
This estimate is calculated by comparing the total amount of shared
bytes in each file before and after the dedupe.
.SH OPTIONS
.SS Common options
\f[CR]files\f[R] can refer to a list of regular files and directories or
be a hyphen (\-) to read them from standard input.
If a directory is specified, all regular files within it will also be
scanned.
Duperemove can also be told to recursively scan directories with the
\f[CR]\-r\f[R] switch.
.TP
\f[B]\-r\f[R]
Enable recursive dir traversal.
.TP
\f[B]\-d\f[R]
De\-dupe the results \- only works on \f[CR]btrfs\f[R] and
\f[CR]xfs\f[R].
Use this option twice to disable the check and try to run the ioctl
anyway.
.TP
\f[B]\-\-hashfile\f[R]=\f[CR]hashfile\f[R]
Use a file for storage of hashes instead of memory.
This option drastically reduces the memory footprint of duperemove and
is recommended when your data set is more than a few files large.
\f[CR]Hashfiles\f[R] are also reusable, allowing you to further reduce
the amount of hashing done on subsequent dedupe runs.
.RS
.PP
If \f[CR]hashfile\f[R] does not exist it will be created.
If it exists, \f[CR]duperemove\f[R] will check the file paths stored
inside of it for changes.
Files which have changed will be rescanned and their updated hashes will
be written to the \f[CR]hashfile\f[R].
Deleted files will be removed from the \f[CR]hashfile\f[R].
.PP
New files are only added to the \f[CR]hashfile\f[R] if they are
discoverable via the \f[CR]files\f[R] argument.
For that reason you probably want to provide the same \f[CR]files\f[R]
list and \f[CR]\-r\f[R] arguments on each run of \f[CR]duperemove\f[R].
The file discovery algorithm is efficient and will only visit each file
once, even if it is already in the \f[CR]hashfile\f[R].
.PP
Adding a new path to a hashfile is as simple as adding it to the
\f[CR]files\f[R] argument.
.PP
When deduping from a hashfile, duperemove will avoid deduping files
which have not changed since the last dedupe.
.RE
.TP
\f[B]\-B\f[R] \f[CR]N\f[R], \f[B]\-\-batchsize\f[R]=\f[CR]N\f[R]
Run the deduplication phase every \f[CR]N\f[R] files newly scanned.
This greatly reduces memory usage for large dataset, or when you are
doing partial extents lookup, but reduces multithreading efficiency.
.RS
.PP
Because of that small overhead, its value shall be selected based on the
average file size and \f[CR]blocksize\f[R].
.PP
The default is a sane value for extents\-only lookups, while you can go
as low as \f[CR]1\f[R] if you are running \f[CR]duperemove\f[R] on very
large files (like virtual machines etc).
.PP
By default, batching is set to 1024.
.RE
.TP
\f[B]\-h\f[R]
Print numbers in human\-readable format.
.TP
\f[B]\-q\f[R]
Quiet mode.
Duperemove will only print errors and a short summary of any dedupe.
.TP
\f[B]\-v\f[R]
Be verbose.
.TP
\f[B]\-\-help\f[R]
Prints help text.
.SS Advanced options
.TP
\f[B]\-\-fdupes\f[R]
Run in \f[CR]fdupes\f[R] mode.
With this option you can pipe the output of \f[CR]fdupes\f[R] to
duperemove to dedupe any duplicate files found.
When receiving a file list in this manner, duperemove will skip the
hashing phase.
.TP
\f[B]\-L\f[R]
Print all files in the hashfile and exit.
Requires the \f[CR]\-\-hashfile\f[R] option.
Will print additional information about each file when run with
\f[CR]\-v\f[R].
.TP
\f[B]\-R\f[R] \f[CR]files ..\f[R]
Remove file from the db and exit.
Duperemove will read the list from standard input if a hyphen (\-) is
provided.
Requires the \f[CR]\-\-hashfile\f[R] option.
.RS
.PP
\f[CR]Note:\f[R] If you are piping filenames from another duperemove
instance it is advisable to do so into a temporary file first as running
duperemove simultaneously on the same hashfile may corrupt that
hashfile.
.RE
.TP
\f[B]\-\-skip\-zeroes\f[R]
Read data blocks and skip any zeroed blocks, useful for speedup
duperemove, but can prevent deduplication of zeroed files.
.TP
\f[B]\-b\f[R] \f[CR]size\f[R]
Use the specified block size for reading file extents.
Defaults to 128K.
.TP
\f[B]\-\-io\-threads\f[R]=\f[CR]N\f[R]
Use N threads for I/O.
This is used by the file hashing and dedupe stages.
Default is automatically detected based on number of host cpus.
.TP
\f[B]\-\-cpu\-threads\f[R]=\f[CR]N\f[R]
Use N threads for CPU bound tasks.
This is used by the duplicate extent finding stage.
Default is automatically detected based on number of host cpus.
.RS
.PP
\f[CR]Note:\f[R] Hyperthreading can adversely affect performance of the
extent finding stage.
If duperemove detects an Intel CPU with hyperthreading it will use half
the number of cores reported by the system for cpu bound tasks.
.RE
.TP
\f[B]\-\-dedupe\-options\f[R]=\f[CR]options\f[R]
Comma separated list of options which alter how we dedupe.
Prepend `no' to an option in order to turn it off.
.RS
.TP
\f[B][no]partial\f[R]
Duperemove can often find more dedupe by comparing portions of extents
to each other.
This can be a lengthy, CPU intensive task so it is turned off by
default.
Using \f[CR]\-\-batchsize\f[R] is recommended to limit the negative
effects of this option.
.RS
.PP
The code behind this option is under active development and as a result
the semantics of the \f[CR]partial\f[R] argument may change.
.RE
.TP
\f[B][no]same\f[R]
Defaults to \f[CR]on\f[R].
Allow dedupe of extents within the same file.
.TP
\f[B][no]only_whole_files\f[R]
Defaults to \f[CR]off\f[R].
Duperemove will only work on full file.
Both extent\-based and block\-based deduplication will be disabled.
The hashfile will be smaller, some operations will be faster, but the
deduplication efficiency will indeed be reduced.
.RE
.TP
\f[B]\-\-read\-hashes\f[R]=\f[CR]hashfile\f[R]
\f[B]This option is primarily for testing\f[R].
See the \f[CR]\-\-hashfile\f[R] option if you want to use hashfiles.
.RS
.PP
Read hashes from a hashfile.
A file list is not required with this option.
Dedupe can be done if duperemove is run from the same base directory as
is stored in the hash file (basically duperemove has to be able to find
the files).
.RE
.TP
\f[B]\-\-write\-hashes\f[R]=\f[CR]hashfile\f[R]
\f[B]This option is primarily for testing\f[R].
See the \f[CR]\-\-hashfile\f[R] option if you want to use hashfiles.
.RS
.PP
Write hashes to a hashfile.
These can be read in at a later date and deduped from.
.RE
.TP
\f[B]\-\-debug\f[R]
Print debug messages, forces \f[CR]\-v\f[R] if selected.
.TP
\f[B]\-\-hash\-threads\f[R]=\f[CR]N\f[R]
Deprecated, see \f[CR]\-\-io\-threads\f[R] above.
.TP
\f[B]\-\-exclude\f[R]=\f[CR]PATTERN\f[R]
You can exclude certain files and folders from the deduplication
process.
This might be benefical for skipping subvolume snapshot mounts, for
instance.
Unless you provide a full path for exclusion, the exclude will be
relative to the current working directory.
Another thing to keep in mind is that shells usually expand glob pattern
so the passed in pattern ought to also be quoted.
Taking everything into consideration the correct way to pass an
exclusion pattern is
\f[CR]duperemove \-\-exclude \[dq]/path/to/dir/file*\[dq] /path/to/dir\f[R]
.SH EXAMPLES
.SS Simple Usage
Dedupe the files in directory /foo, recurse into all subdirectories.
You only want to use this for small data sets:
.IP
.EX
duperemove \-dr /foo
.EE
.PP
Use duperemove with fdupes to dedupe identical files below directory
foo:
.IP
.EX
fdupes \-r /foo | duperemove \-\-fdupes
.EE
.SS Using Hashfiles
Duperemove can optionally store the hashes it calculates in a hashfile.
Hashfiles have two primary advantages \- memory usage and re\-usability.
When using a hashfile, duperemove will stream computed hashes to it,
instead of main memory.
.PP
If Duperemove is run with an existing hashfile, it will only scan those
files which have changed since the last time the hashfile was updated.
The \f[CR]files\f[R] argument controls which directories duperemove will
scan for newly added files.
In the simplest usage, you rerun duperemove with the same parameters and
it will only scan changed or newly added files \- see the first example
below.
.PP
Dedupe the files in directory foo, storing hashes in foo.hash.
We can run this command multiple times and duperemove will only checksum
and dedupe changed or newly added files:
.IP
.EX
duperemove \-dr \-\-hashfile=foo.hash foo/
.EE
.PP
Don\[cq]t scan for new files, only update changed or deleted files, then
dedupe:
.IP
.EX
duperemove \-dr \-\-hashfile=foo.hash
.EE
.PP
Add directory bar to our hashfile and discover any files that were
recently added to foo:
.IP
.EX
duperemove \-dr \-\-hashfile=foo.hash foo/ bar/
.EE
.PP
List the files tracked by foo.hash:
.IP
.EX
duperemove \-L \-\-hashfile=foo.hash
.EE
.SH FAQ
.SS Is duperemove safe for my data?
Yes.
To be specific, duperemove does not deduplicate the data itself.
It simply finds candidates for dedupe and submits them to the Linux
kernel FIDEDUPERANGE ioctl.
In order to ensure data integrity, the kernel locks out other access to
the file and does a byte\-by\-byte compare before proceeding with the
dedupe.
.SS Is is safe to interrupt the program (Ctrl\-C)?
Yes.
The Linux kernel deals with the actual data.
On Duperemove\[cq] side, a transactional database engine is used.
The result is that you should be able to ctrl\-c the program at any
point and re\-run without experiencing corruption of your hashfile.
In case of a bug, your hashfile may be broken, but your data never will.
.SS I got two identical files, why are they not deduped?
Duperemove by default works on extent granularity.
What this means is if there are two files which are logically identical
(have the same content) but are laid out on disk with different extent
structure they won\[cq]t be deduped.
For example if 2 files are 128k each and their content are identical but
one of them consists of a single 128k extent and the other of 2 * 64k
extents then they won\[cq]t be deduped.
This behavior is dependent on the current implementation and is subject
to change as duperemove is being improved.
.SS What is the cost of deduplication?
Deduplication will lead to increased fragmentation.
The blocksize chosen can have an effect on this.
Larger blocksizes will fragment less but may not save you as much space.
Conversely, smaller block sizes may save more space at the cost of
increased fragmentation.
.SS How can I find out my space savings after a dedupe?
Duperemove will print out an estimate of the saved space after a dedupe
operation for you.
.PP
You can get a more accurate picture by running `btrfs fi df' before and
after each duperemove run.
.PP
Be careful about using the `df' tool on btrfs \- it is common for space
reporting to be `behind' while delayed updates get processed, so an
immediate df after deduping might not show any savings.
.SS Why is the total deduped data report an estimate?
At the moment duperemove can detect that some underlying extents are
shared with other files, but it can not resolve which files those
extents are shared with.
.PP
Imagine duperemove is examining a series of files and it notes a shared
data region in one of them.
That data could be shared with a file outside of the series.
Since duperemove can\[cq]t resolve that information it will account the
shared data against our dedupe operation while in reality, the kernel
might deduplicate it further for us.
.SS Why are my files showing dedupe but my disk space is not shrinking?
This is a little complicated, but it comes down to a feature in Btrfs
called \f[I]bookending\f[R].
The \c

Changelog.md
PENDING
---

**Notable changes**:


Version 0.15
---

**Notable changes**:
 * Files are no longer invalidated when renamed.
 * The scan phase now reports both files and bytes in its progress, which is especially useful when processing large files.

Version 0.14
---

**Notable changes**:
 * Batching has been reimplemented on top of the dedupe\_seq.
 * The "scan" phase has been reimplemented (see 8264336ea2a3b78e3bdce162fc389d02338af326 for details).
 * Filesystem locking has been implemented. See f3947e9606f103417537974bc3dda4f6254c4503 for details.

Version 0.13
---

**Notable changes**:
 * Add a new dedupe option: `[no]rescan_files`. It will increase performance in some use cases.
 * New behaviors from v0.12 has been consolidated. Extent-based lookup is always enabled, as is fiemap. The v2 hashfile is no longer supported.
 * Hashfile are now updated after deduplication, to reflect the new physical offsets. This avoid (re)deduplicating extents in some cases.
 * Partial mode has been enhanced to support batching. The overall performance of this mode (which was previously known as "block-based mode") has been improved.
 * All files are now open in readonly mode.
 * Hashfile version has been increased to reflect the new database behaviors. Previous hashfiles are not compatible.
 * Always compute a hash for the entire file. This let us deduplicate same files easily, regardless of their extents mappings.
 * Deduplicating only parts of a file can be disabled using the `[no]only_whole_files` dedupe option.
 * Hashfiles with unsupported features or hash algorithm are now recreated transparently. Migration of the old content is not implemented.
 * Relative exclude patterns are no longer silently ingested. Such patterns are now rebuilt on top of the current working directory.
 * Batching is now set to 1024 by default.

Version 0.12
---

**Notable changes**:
 * Duplication lookup is now based on extents. This leads to a massive increase of the performances. Block-based lookup is still possible via `--dedupe-options=partial`.
 * Following that change, a new hashfile format has been introduced. Previous hashfile format is still supported when extents lookup are disabled, this is not recommended.
 * Batching has been implemented. When enabled with the `-B <batchsize>` option, `duperemove` will run the deduplication phase every `<batchsize>` scanned files. This is meant to help running `duperemove` on large dataset, with small blocksize, or on memory-constrained systems.
 * All hash algorithms has been removed and replaced by xxh128. This variant is as robust as murmur3 while being faster. Choosing a hash function via the `--hash` option has been removed. Hashfiles built with other algorithm must be removed.


duperemove.c
/*
 * duperemove.c
 *
 * Copyright (C) 2013 SUSE.  All rights reserved.
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public
 * License version 2 as published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * Authors: Mark Fasheh <mfasheh@suse.de>
 */

#include <sys/types.h>
#include <fcntl.h>
#include <unistd.h>
#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <stddef.h>
#include <errno.h>
#include <string.h>
#include <getopt.h>
#include <inttypes.h>

#include <glib.h>

#include "list.h"
#include "csum.h"
#include "filerec.h"
#include "hash-tree.h"
#include "results-tree.h"
#include "dedupe.h"
#include "util.h"
#include "btrfs-util.h"
#include "dbfile.h"
#include "memstats.h"
#include "debug.h"
#include "progress.h"
#include "file_scan.h"
#include "find_dupes.h"
#include "run_dedupe.h"

#include "opt.h"

unsigned int blocksize = DEFAULT_BLOCKSIZE;

static int stdin_filelist = 0;
static unsigned int list_only_opt = 0;
static unsigned int rm_only_opt = 0;
struct dbfile_config dbfile_cfg;

static enum {
	H_READ,
	H_WRITE,
	H_UPDATE,
} use_hashfile = H_UPDATE;

static void print_file(char *filename, char *ino, char *subvol)
{
	if (verbose)
		printf("%s\t%s\t%s\n", filename, ino, subvol);
	else
		printf("%s\n", filename);
}

static int list_db_files(char *filename)
{
	int ret;

	_cleanup_(sqlite3_close_cleanup) struct dbhandle *db = dbfile_open_handle(filename);
	if (!db) {
		eprintf("Error: Could not open \"%s\"\n", filename);
		return -1;
	}

	ret = dbfile_iter_files(db, &print_file);
	return ret;
}

static void rm_db_files_from_stdin(struct dbhandle *db)
{
	_cleanup_(freep) char *path = NULL;
	size_t pathlen = 0;
	ssize_t readlen;

	while ((readlen = getline(&path, &pathlen, stdin)) != -1) {
		if (readlen == 0)
			continue;

		if (readlen > 0 && path[readlen - 1] == '\n') {
			path[--readlen] = '\0';
		}

		if (readlen > PATH_MAX - 1) {
			eprintf("Path max exceeded: %s\n", path);
			continue;
		}

		dbfile_remove_file(db, path);
	}
}

static int rm_db_files(int numfiles, char **files)
{
	int i, ret;
	_cleanup_(sqlite3_close_cleanup) struct dbhandle *db = dbfile_open_handle(options.hashfile);
	if (!db) {
		eprintf("Error: Could not open \"%s\"\n", options.hashfile);
		return -1;
	}

	for (i = 0; i < numfiles; i++) {
		const char *name = files[i];

		if (strlen(name) == 1 && name[0] == '-')
			rm_db_files_from_stdin(db);

		ret = dbfile_remove_file(db, name);
		if (ret == 0)
			vprintf("Removed \"%s\" from hashfile.\n", name);

		if (ret)
			printf("ret ?\n");
	}
	return 0;
}

static void print_version(void)
{
	char *s = NULL;
#ifdef	DEBUG_BUILD
	s = " (debug build)";
#endif
	printf("duperemove %s%s\n", VERSTRING, s ? s : "");
}

/* adapted from ocfs2-tools */
static int parse_dedupe_opts(const char *raw_opts)
{
	_cleanup_(freep) char *opts;
	char *token, *next, *p, *arg;
	int print_usage = false;
	int invert, ret = 0;

	opts = strdup(raw_opts);

	for (token = opts; token && *token; token = next) {
		p = strchr(token, ',');
		next = NULL;
		invert = 0;

		if (p) {
			*p = '\0';
			next = p + 1;
		}

		arg = strstr(token, "no");
		if (arg == token) {
			invert = 1;
			token += strlen("no");
		}

		if (strcmp(token, "same") == 0) {
			options.dedupe_same_file = !invert;
		} else if (strcmp(token, "partial") == 0) {
			options.do_block_hash = !invert;
		} else if (strcmp(token, "only_whole_files") == 0) {
			options.only_whole_files = !invert;
		} else {
			print_usage = true;
			break;
		}
	}

	if (print_usage) {
		eprintf("Bad dedupe options specified. Valid dedupe "
			"options are:\n"
			"\t[no]same\n"
			"\t[no]only_whole_files\n"
			"\t[no]partial\n");
		ret = EINVAL;
	}

	return ret;
}

enum {
	DEBUG_OPTION = CHAR_MAX + 1,
	HELP_OPTION,
	VERSION_OPTION,
	WRITE_HASHES_OPTION,
	READ_HASHES_OPTION,
	HASHFILE_OPTION,
	IO_THREADS_OPTION,
	CPU_THREADS_OPTION,
	SKIP_ZEROES_OPTION,
	FDUPES_OPTION,
	DEDUPE_OPTS_OPTION,
	QUIET_OPTION,
	EXCLUDE_OPTION,
	BATCH_SIZE_OPTION,
};

static int process_fdupes(void)
{
	int ret = 0;
	_cleanup_(freep) char *path = NULL;
	size_t pathlen = 0;
	ssize_t readlen;

	while ((readlen = getline(&path, &pathlen, stdin)) != -1) {
		if (readlen == 0)
			continue;

		if (readlen == 1 && path[0] == '\n') {
			ret = fdupes_dedupe();
			if (ret)
				return ret;
			free_all_filerecs();
			continue;
		}

		if (readlen > 0 && path[readlen - 1] == '\n') {
			path[--readlen] = '\0';
		}

		if (readlen > PATH_MAX - 1) {
			eprintf("Path max exceeded: %s\n", path);
			continue;
		}

		add_file_fdupes(path);
	}

	return 0;
}

static int add_files_from_stdin(struct dbhandle *db)
{
	_cleanup_(freep) char *path = NULL;
	size_t pathlen = 0;
	ssize_t readlen;

	while ((readlen = getline(&path, &pathlen, stdin)) != -1) {
		if (readlen == 0)
			continue;

		if (readlen > 0 && path[readlen - 1] == '\n') {
			path[--readlen] = '\0';
		}

		if (readlen > PATH_MAX - 1) {
			eprintf("Path max exceeded: %s\n", path);
			continue;
		}

		if (scan_file(path, db)) {
			eprintf("Error: cannot add %s into the lookup list\n",
				path);
			return 1;
		}
	}

	return 0;
}

static int scan_files_from_cmdline(int numfiles, char **files, struct dbhandle *db)
{
	int i;

	for (i = 0; i < numfiles; i++) {
		char *name = files[i];

		if (scan_file(name, db)) {
			eprintf("Error: cannot scan %s\n", name);
			return 1;
		}
	}

	return 0;
}

static void help(void)
{
	execlp("man", "man", "8", "duperemove", NULL);
}

/*
 * Ok this is doing more than just parsing options.
 */
static int parse_options(int argc, char **argv, int *filelist_idx)
{
	int c, numfiles;
	bool read_hashes = false;
	bool write_hashes = false;
	bool update_hashes = false;

	static struct option long_ops[] = {
		{ "debug", 0, NULL, DEBUG_OPTION },
		{ "help", 0, NULL, HELP_OPTION },
		{ "version", 0, NULL, VERSION_OPTION },
		{ "write-hashes", 1, NULL, WRITE_HASHES_OPTION },
		{ "read-hashes", 1, NULL, READ_HASHES_OPTION },
		{ "hashfile", 1, NULL, HASHFILE_OPTION },
		{ "io-threads", 1, NULL, IO_THREADS_OPTION },
		{ "hash-threads", 1, NULL, IO_THREADS_OPTION },
		{ "cpu-threads", 1, NULL, CPU_THREADS_OPTION },
		{ "skip-zeroes", 0, NULL, SKIP_ZEROES_OPTION },
		{ "fdupes", 0, NULL, FDUPES_OPTION },
		{ "dedupe-options=", 1, NULL, DEDUPE_OPTS_OPTION },
		{ "quiet", 0, NULL, QUIET_OPTION },
		{ "exclude", 1, NULL, EXCLUDE_OPTION },
		{ "batchsize", 1, NULL, BATCH_SIZE_OPTION },
		{ NULL, 0, NULL, 0}
	};

	if (argc < 2) {
		help(); /* Never returns */
	}

	while ((c = getopt_long(argc, argv, "b:vdDrh?LRqB:", long_ops, NULL))
	       != -1) {
		switch (c) {
		case 'b':
			blocksize = parse_size(optarg);
			if (blocksize < MIN_BLOCKSIZE ||
			    blocksize > MAX_BLOCKSIZE){
				eprintf("Error: Blocksize is bounded by %u and %u, %u found\n",
					MIN_BLOCKSIZE, MAX_BLOCKSIZE, blocksize);
				return EINVAL;
			}
			break;
		case 'd':
		case 'D':
			options.run_dedupe += 1;
			break;
		case 'r':
			options.recurse_dirs = true;
			break;
		case VERSION_OPTION:
			print_version();
			exit(0);
		case DEBUG_OPTION:
			debug = 1;
			verbose = 1;
			break;
		case 'v':
			verbose = 1;
			break;
		case 'h':
			human_readable = 1;
			break;
		case WRITE_HASHES_OPTION:
			write_hashes = true;
			options.hashfile = strdup(optarg);
			break;
		case READ_HASHES_OPTION:
			read_hashes = true;
			options.hashfile = strdup(optarg);
			break;
		case HASHFILE_OPTION:
			update_hashes = true;
			options.hashfile = strdup(optarg);
			break;
		case IO_THREADS_OPTION:
			options.io_threads = strtoul(optarg, NULL, 10);
			if (!options.io_threads){
				eprintf("Error: --io-threads must be "
					"an integer, %s found\n", optarg);
				return EINVAL;
			}
			break;
		case CPU_THREADS_OPTION:
			options.cpu_threads = strtoul(optarg, NULL, 10);
			if (!options.cpu_threads){
				eprintf("Error: --cpu-threads must be "
					"an integer, %s found\n", optarg);
				return EINVAL;
			}
			break;
		case SKIP_ZEROES_OPTION:
			options.skip_zeroes = true;
			break;
		case FDUPES_OPTION:
			options.fdupes_mode = 1;
			break;
		case DEDUPE_OPTS_OPTION:
			if (parse_dedupe_opts(optarg))
				return EINVAL;
			break;
		case 'L':
			list_only_opt = 1;
			break;
		case 'R':
			rm_only_opt = 1;
			break;
		case QUIET_OPTION:
		case 'q':
			quiet = 1;
			break;
		case EXCLUDE_OPTION:
			if (add_exclude_pattern(optarg))
				eprintf("Error: cannot exclude %s\n", optarg);
			break;
		case BATCH_SIZE_OPTION:
		case 'B':
			options.batch_size = parse_size(optarg);
			break;
		case HELP_OPTION:
			help();
			break;
		case '?':
		default:
			return 1;
		}
	}

	numfiles = argc - optind;

	if (options.only_whole_files && options.do_block_hash) {
		eprintf("Error: using both only_whole_files and partial "
			"options have no meaning\n");
		return 1;
	}

	/* Filter out option combinations that don't make sense. */
	if ((write_hashes + read_hashes + update_hashes) > 1) {
		eprintf("Error: Specify only one hashfile option.\n");
		return 1;
	}

	if (read_hashes)
		use_hashfile = H_READ;
	else if (write_hashes)
		use_hashfile = H_WRITE;
	else if (update_hashes)
		use_hashfile = H_UPDATE;

	/*
	 * Always add the hashfile and its wal etc to the exclude list
	 * A wildcard would be easier but may exclude extra files silently,
	 * this would be confusing for the user.
	 */
	if (options.hashfile != NULL) {
		char tmp[PATH_MAX + 10 ] = {0,};
		add_exclude_pattern(options.hashfile);
		snprintf(tmp, PATH_MAX + 9, "%s-wal", options.hashfile);
		add_exclude_pattern(tmp);
		snprintf(tmp, PATH_MAX + 9, "%s-shm", options.hashfile);
		add_exclude_pattern(tmp);
	}

	if (read_hashes) {
		if (numfiles) {
			eprintf("Error: --read-hashes option does not take a "
				"file list argument\n");
			return 1;
		}
		goto out_nofiles;
	}

	if (options.fdupes_mode) {
		if (read_hashes || write_hashes || update_hashes) {
			eprintf("Error: cannot mix hashfile option with "
				"--fdupes option\n");
			return 1;
		}

		if (numfiles) {
			eprintf("Error: fdupes option does not take a file "
				"list argument\n");
			return 1;
		}
		/* rest of fdupes mode is implemented in main() */
		return 0;
	}

	*filelist_idx = optind;
	if (numfiles == 1 && strcmp(argv[optind], "-") == 0)
		stdin_filelist = 1;

	if (list_only_opt && rm_only_opt) {
		eprintf("Error: Can not mix '-L' and '-R' options.\n");
		return 1;
	}

	if (list_only_opt || rm_only_opt) {
		if (!options.hashfile || use_hashfile == H_WRITE) {
			eprintf("Error: --hashfile= option is required "
				"with '-L' or -R.\n");
			return 1;
		}

		if (list_only_opt && numfiles) {
			eprintf("Error: -L option do not take "
				"a file list argument\n");
			return 1;
		}
	}

	if (!(options.fdupes_mode || list_only_opt)
			&& numfiles == 0) {
		eprintf("Error: a file list argument is required.\n");
		return 1;
	}

out_nofiles:
	return 0;
}

static void print_header(void)
{
	vprintf("Using %uK blocks\n", blocksize / 1024);
	vprintf("Using %s hashing\n", options.do_block_hash ? "block+extent" : "extent");
#ifdef	DEBUG_BUILD
	printf("Debug build, performance may be impacted.\n");
#endif
	qprintf("Gathering file list...\n");
}

static void __process_duplicates(struct dbhandle *db, unsigned int seq)
{
	int ret;
	struct results_tree res;
	struct hash_tree dups_tree;

	init_results_tree(&res);
	init_hash_tree(&dups_tree);

	qprintf("Loading only identical files from hashfile.\n");
	ret = dbfile_load_same_files(db, &res, seq + 1);
	if (ret)
		goto out;

	if (options.run_dedupe)
		dedupe_results(&res, true);
	else
		print_dupes_table(&res, true);

	/* Reset the results_tree before loading extents or blocks */
	free_results_tree(&res);

	if (!options.only_whole_files) {
		init_results_tree(&res);

		qprintf("Loading only duplicated hashes from hashfile.\n");

		ret = dbfile_load_extent_hashes(db, &res, seq + 1);
		if (ret)
			goto out;

		printf("Found %llu identical extents.\n", res.num_extents);
		if (options.do_block_hash) {
			ret = dbfile_load_block_hashes(db, &dups_tree, seq + 1);
			if (ret)
				goto out;

			ret = find_additional_dedupe(&res);
			if (ret)
				goto out;
		}

		if (options.run_dedupe)
			dedupe_results(&res, false);
		else
			print_dupes_table(&res, false);
	}

out:
	free_results_tree(&res);
	free_hash_tree(&dups_tree);
}

static void process_duplicates(struct dbhandle *db)
{
	unsigned int max = get_max_dedupe_seq(db);

	/* Spawn a dedicated thread pool to block-based lookup */
	if (options.do_block_hash)
		extents_search_init();

	for (unsigned int i = dedupe_seq; i < max; i++) {
		/* Drop all filerecs from the previous iteration. Needed filerecs will be
		 * recreated by __process_duplicates()
		 */
		free_all_filerecs();
		__process_duplicates(db, i);

		if (options.run_dedupe) {
			/*
			 * Bump dedupe_seq, this effectively marks the files
			 * in our hashfile as having been through dedupe.
			 */
			dedupe_seq++;
			dbfile_cfg.dedupe_seq = dedupe_seq;
			dbfile_cfg.blocksize = blocksize;
			dbfile_sync_config(db, &dbfile_cfg);
		}
	}

	if (options.do_block_hash)
		extents_search_free();
}

static int scan_files(int argc, char **argv, int filelist_idx, struct dbhandle *db)
{
	int ret;

	filescan_init();
	if (!quiet)
		pscan_run();

	if (stdin_filelist)
		ret = add_files_from_stdin(db);
	else
		ret = scan_files_from_cmdline(argc - filelist_idx,
					     &argv[filelist_idx], db);

	pscan_finish_listing();
	filescan_free();
	if (!quiet)
		pscan_join();

	if (ret)
		return ret;

	/*
	 * Sync the locked filesystem informations in the hashfile
	 */
	fs_get_locked_uuid(&(dbfile_cfg.fs_uuid));
	ret = dbfile_sync_config(db, &dbfile_cfg);
	if (ret)
		return ret;

	return 0;
}

int main(int argc, char **argv)
{
	int ret, filelist_idx = 0;
	_cleanup_(sqlite3_close_cleanup) struct dbhandle *db = NULL;

	char stdbuf[BUFSIZ];
	setvbuf(stdout, stdbuf, _IOLBF, BUFSIZ);

	init_filerec();

	/* Set the default CPU limits before parsing the user options */
	get_num_cpus(&(options.cpu_threads), &(options.io_threads));

	ret = parse_options(argc, argv, &filelist_idx);
	if (ret) {
		exit(1);
	}

	/* Allow larger than unusal amount of open files. On linux
	 * this should bw increase form 1K to 512K open files
	 * simultaneously.
	 *
	 * On multicore SSD machines it's not hard to get to 1K open
	 * files.
	 */
	increase_limits();

	if (options.fdupes_mode)
		return process_fdupes();

	if (list_only_opt)
		return list_db_files(options.hashfile);
	else if (rm_only_opt)
		return rm_db_files(argc - filelist_idx, &argv[filelist_idx]);

	db = dbfile_open_handle(options.hashfile);
	if (!db)
		goto out;

	dbfile_set_gdb(db);

	ret = dbfile_get_config(db->db, &dbfile_cfg);
	if (ret)
		goto out;

	dedupe_seq = dbfile_cfg.dedupe_seq;

	print_header();

	if (use_hashfile == H_WRITE || use_hashfile == H_UPDATE) {
		ret = dbfile_prune_unscanned_files(db);
		if (ret) {
			eprintf("Unable to prune unscanned files\n");
			goto out;
		}

		ret = scan_files(argc, argv, filelist_idx, db);
		if (ret)
			goto out;

		qprintf("Hashfile \"%s\" written\n",
			options.hashfile);
	}

	if (use_hashfile == H_READ || use_hashfile == H_UPDATE)
		process_duplicates(db);

out:
	free_all_filerecs();

#ifdef DEBUG_BUILD
	print_mem_stats();
#else
	if (ret == ENOMEM || debug)
		print_mem_stats();
#endif

	return ret;
}
